forge-game/src/main/java/forge/game/phase/PhaseHandler.java

```diff
@@ -1023,87 +1025,108 @@ public void startFirstTurn(Player goesFirst, Runnable startGameHook) {
+                System.out.println("Active player is no longer in the game...");
+                pPlayerPriority = game.getNextPlayerAfter(getPriorityPlayer());
+                pFirstPriority = pPlayerPriority;
+            }
 
-                        saHost = game.getCardState(saHost);
-                        final Zone currentZone = saHost.getZone();
+            if (chosenSa != null) {
+                if (DEBUG_PHASES) {
+                    System.out.print("... " + pPlayerPriority + " plays " + chosenSa);
+                }
+                for (SpellAbility sa : chosenSa) {
+                    Card saHost = sa.getHostCard();
+                    final Zone originZone = saHost.getZone();
 
-                        // Need to check if Zone did change
-                        if (currentZone != null && originZone != null && !currentZone.equals(originZone) && (sa.isSpell() || sa instanceof LandAbility)) {
-                            // currently there can be only one Spell put on the Stack at once, or Land Abilities be played
-                            final CardZoneTable triggerList = new CardZoneTable();
-                            triggerList.put(originZone.getZoneType(), currentZone.getZoneType(), saHost);
-                            triggerList.triggerChangesZoneAll(game, sa);
-                        }
+                    // Note: this involves going back to the controller (player or AI) for choices
+                    if (pPlayerPriority.getController().playChosenSpellAbility(sa)) {
+                        pFirstPriority = pPlayerPriority; // all opponents have to pass before stack is allowed to resolve
                     }
-                    game.copyLastState();
-                    loopCount++;
-                } while (loopCount < 999 || !pPlayerPriority.getController().isAI());
 
+                    saHost = game.getCardState(saHost);
+                    final Zone currentZone = saHost.getZone();
+
+                    // Need to check if Zone did change
+                    if (currentZone != null && originZone != null && !currentZone.equals(originZone) && (sa.isSpell() || sa instanceof LandAbility)) {
+                        // currently there can be only one Spell put on the Stack at once, or Land Abilities be played
+                        final CardZoneTable triggerList = new CardZoneTable();
+                        triggerList.put(originZone.getZoneType(), currentZone.getZoneType(), saHost);
+                        triggerList.triggerChangesZoneAll(game, sa);
+                    }
+                }
+                game.copyLastState();
+                loopCount++;
                 if (loopCount >= 999 && pPlayerPriority.getController().isAI()) {
                     System.out.print("AI looped too much with: " + chosenSa);
                 }
+                keep_looping = loopCount < 999 || !pPlayerPriority.getController().isAI();
```

Comment:
I don't really understand why this is there, I just copied it from the previous implementation.

Reply:
Yea i know its just to prevent the AI from over looping when its making decisions. I was just saying if we pulled the numeric up to the top of the file (here and the line just above it) then we can alter the "looped too many times" logic easier.